modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ListenerHelper.java

```diff
@@ -0,0 +1,508 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+// Original code is re-licensed to Oracle by the author.
+// https://github.com/andy-goryachev/FxTextEditor/blob/master/src/goryachev/fx/FxDisconnector.java
+// Copyright © 2021-2022 Andy Goryachev <andy@goryachev.com>
+package com.sun.javafx.scene.control;
+
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import javafx.beans.InvalidationListener;
+import javafx.beans.Observable;
+import javafx.beans.value.ChangeListener;
+import javafx.beans.value.ObservableValue;
+import javafx.collections.ListChangeListener;
+import javafx.collections.MapChangeListener;
+import javafx.collections.ObservableList;
+import javafx.collections.ObservableMap;
+import javafx.collections.ObservableSet;
+import javafx.collections.SetChangeListener;
+import javafx.concurrent.Task;
+import javafx.event.Event;
+import javafx.event.EventHandler;
+import javafx.event.EventType;
+import javafx.scene.Node;
+import javafx.scene.Scene;
+import javafx.scene.control.MenuItem;
+import javafx.scene.control.SkinBase;
+import javafx.scene.control.TableColumnBase;
+import javafx.scene.control.TreeItem;
+import javafx.scene.transform.Transform;
+import javafx.stage.Window;
+
+/**
+ * This class provides convenience methods for adding various listeners, both
+ * strong and weak, as well as a single {@link #disconnect()} method to remove
+ * all listeners.
+ * <p>
+ * There are two usage patterns:
+ * <ul>
+ * <li>Client code registers a number of listeners and removes them all at once
+ * via {@link #disconnect()} call.
+ * <li>Client code registers a number of listeners and removes one via its
+ * {@link IDisconnectable} instance.
+ * </ul>
+ */
+public class ListenerHelper implements IDisconnectable {
```

Comment:
you are correct: the original intent for this class was to make it a general purpose facility to help with listeners, something that might be useful at the application level (and it used `ListenerHelper.get(Node)`).  Since that would require CSR and a public discussion, we decided to hide it as an implementation detail in skins (to unblock skin memory leak fixes).

Once we go through all the necessary discussions and everyone agrees, we could introduce it as a public API in the javafx.base module.

Reply:
Exactly. In its current form, `ListenerHelper` is a utility helper class for Skins, and should be reviewed with that in mind. As discussed in previous comments, there are several things that will need to change when/if this is proposed as a more general utility. We can defer that discuss, since this is entirely an internal helper class for now.

As a next step, after this is integrated and before any discussion of making this a public utility, it can be used as a replacement for `LambdaMultiplePropertyChangeListenerHandler` -- see [JDK-8296076](https://bugs.openjdk.org/browse/JDK-8296076).

Reply:
Although it is an internal (non-public) utility class, once introduced, it will be tempting to use it from non-Skin code as well.
If the current scope is limited to provide "a utility helper class for Skins" then how about doing one of two things-
1) Mention the current intention and possible future modifications as a block comment above the `ListenerHelper` class
2) Keep only common methods in`ListenerHelper` class and move `Skin` specific code to a class derived from `ListenerHelper`. Skins should start using this derived class.